Fix/merge wiki content into page

ulferts opened the pull request:
Merges the WikiContent model into the WikiPage model. Presumably, the two were separate to support multiple text version of the same page but that reason was no longer relevant with the introduction of the journals. Nowadays, having the two separated does not grant any benefits while increasing complexity.

https://community.openproject.org/projects/openproject/work_packages/32796

The PR does not attempt to alter the behaviour of application in any way. 

For the most part, the changes are just the removals of indirections (referencing the wiki page instead of the wiki page's content). In the frontend however a specific view object (`WikiPages::AtVersion`) has been introduced. Some views, most notably the show pages allow for browsing through different versions of the text. Instead of having multiple objects referenced in the view, the `WikiPages::AtVersion` object is passed, representing a wiki page at the version requested by the user. Since this affects the whole page, using ViewComponents has not seemed like an appropriate fit. There is no other view component as of now yet which is why the class is placed under `app/helpers`.

Limitations existing before have been kept. E.g. the journals only cover changes in the text and not in the title or the parent page. The `author` field existing both in the page (used to be content) as well as in the journal has also been kept. They arguable are unnecessary since the user reference is kept in the journal anyway. 

## TODO

* [x] Move code from wiki_content.rb into wiki_page.rb
* [x] Write migration to
  * alter the table structure of both wiki_page as well as rename wiki_content_journals into wiki_page_journals. Does not lead to more attributes being journalized but allows to later include e.g. the title or the parent page in the journalization. 
  * migrate the existing data
* [x] Adapt the services
* [x] Adapt the contracts
* [x] Adapt the controller
* [x] Adapt the views  

ulferts committed:
merge wiki_content into page

dombesz commented on app/views/wiki/show.html.erb:
```diff
@@ -28,38 +28,35 @@ See COPYRIGHT and LICENSE files for more details.
+      </li>
     <% end %>
-    <% if @content.version == @page.content.version %>
+    <% if @page.current_version? %>
```
I tried to request non-existing version and on the dev branch it returns 1 version, without the next/prev buttons.
<img width="1280" alt="image" src="https://user-images.githubusercontent.com/83396/234326089-cdbce73c-2a9b-4a65-b160-252ce920f293.png">

But on this branch it displays the prev/next buttons and it shows inconsistent version numbers 6/1.

<img width="1277" alt="image" src="https://user-images.githubusercontent.com/83396/234331254-952a23c8-db16-472d-9d8c-2e9d91cd7df7.png">

dombesz commented on app/controllers/wiki_controller.rb:
```diff
@@ -109,34 +108,31 @@ def new_child
     end
 
-    @content = @page.content_for_version(params[:version])
+    version = params[:version] if User.current.allowed_to?(:view_wiki_edits, @project)
```
Editing a certain version always brings the latest version of the page, regardless of what version I try to edit.

<img width="1280" alt="image" src="https://user-images.githubusercontent.com/83396/234348640-3b789ddc-06dd-43f7-b362-ca572991d7a9.png">

This is the content of v1

<img width="1098" alt="image" src="https://user-images.githubusercontent.com/83396/234349438-2c1e3505-2ac0-4ed0-8f82-df3bfa44384c.png">


<img width="1280" alt="image" src="https://user-images.githubusercontent.com/83396/234348702-4ba024f5-f8d3-4225-b703-2d6ed421609c.png">

<img width="1280" alt="image" src="https://user-images.githubusercontent.com/83396/234348753-404af443-a106-4683-b97b-04f0536a8a97.png">

<img width="1106" alt="image" src="https://user-images.githubusercontent.com/83396/234348962-d6196acd-7d66-40e8-8ba3-af9b5141d360.png">

This is the history:

<img width="1098" alt="image" src="https://user-images.githubusercontent.com/83396/234348876-bb7b2f3f-d5b8-4b1b-9a67-9085a0c989ad.png">

I can also edit a version that was created by another user via the link, but the edit button is not visible.

<img width="1085" alt="image" src="https://user-images.githubusercontent.com/83396/234349156-6169e73f-feb0-4e46-9540-21addc317d8b.png">

ulferts replied:
@dombesz good catch. It took me a while to figure out what was going on. `form_for` will call `to_model` if that is defined in the object passed which it was because of the `SimpleDelegator` ancestor in the `AtVersion` class. But that lead to the wrapped model being returned which always had the most recent text. I prevented this to happen now by no longer responding to `to_model`. 

I then also went ahead and simplified the form structure. With the distinction between page and content gone, that was now possible but it of course lead to other errors. But now, it should be good.

dombesz commented on app/services/projects/copy/wiki_dependent_service.rb:
```diff
@@ -69,9 +69,6 @@ def copy_wiki_pages(_params)
 
     def copy_wiki_page(source_page, new_parent_id)
-      # Skip pages without content
-      return if source_page.content.nil?
```
Should we have `return if source_page.text.nil?` instead? Admittedly I think, it makes no sense to not copy empty pages, so I'm fine with not having the guard clause here.

ulferts replied:
I think the level this guard was on was data consistence. So it did not guard against the page not having text (then it would have to have checked `source_page.content.text.present?`). Rather, it checked that there was a `content` object associated to the page. 

Such a case can no longer occur so I think it safe to simply remove the guard.

dombesz reviewed (changes requested):
Looks like a nice refactoring on the wiki pages, congrats @ulferts !🎉
I left a few findings with screenshots.

ulferts committed:
Merge branch 'dev' into fix/merge_wiki_content_into_page

ulferts committed:
guard against illegal wiki page versions

ulferts committed:
enable editing former versions of wiki page

This is done, so former versions can be rolled back to. The transformation of the view object to it`s wrapped model needs to be prevented.

Once that had been achieved more indirections via content were removable.

ulferts committed:
fix seeding reference

ulferts committed:
Merge remote-tracking branch 'origin/dev' into fix/merge_wiki_content_into_page

ulferts committed:
rely on delegation again

dombesz reviewed (approved):
LGTM! 🎉
Feel free to merge once CI is green.

ulferts committed:
adapt project seeder spec

estan commented:
@ulferts In attempting to upgrade our instance to 13, we're getting

```
seeder_1    | PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_journals_on_journable_type_and_journable_id_and_version"
seeder_1    | DETAIL:  Key (journable_type, journable_id, version)=(WikiPage, 1, 1) already exists.
```

during the `db/migrate/20230322135932_merge_wiki_content_into_page.rb` migration added in this PR.

(When trying to google this problem I came across an old PR that may or may not be relevant: https://github.com/opf/openproject/pull/9598)

estan commented:
@ulferts => https://community.openproject.org/projects/openproject/work_packages/49761

estan commented:
Apologies, I see this is likely fixed with https://github.com/opf/openproject/commit/1a640c2417cb6cea7284dc76bfc5ef6fea617b87 already.

EDIT: Actually no, that commit is in v13.0.0, so that's not a fix, or the fix is not working somehow.
